src/Algebra/Graph/Relation/Symmetric.hs

```diff
@@ -11,27 +11,485 @@
+replaceVertex u v = gmap $ \w -> if w == u then v else w
+
+-- | Merge vertices satisfying a given predicate into a given vertex.
+-- Complexity: /O((n + m) * log(n))/ time, assuming that the predicate takes
+-- /O(1)/ to be evaluated.
+--
+-- @
+-- mergeVertices ('const' False) x    == id
+-- mergeVertices (== x) y             == 'replaceVertex' x y
+-- mergeVertices 'even' 1 (0 * 2)     == 1 * 1
+-- mergeVertices 'odd'  1 (3 + 4 * 5) == 4 * 1
+-- @
+mergeVertices :: Ord a => (a -> Bool) -> a -> SymmetricRelation a -> SymmetricRelation a
+mergeVertices p v = gmap $ \u -> if p u then v else u
+
+-- | Transform a graph by applying a function to each of its vertices. This is
+-- similar to @Functor@'s 'fmap' but can be used with non-fully-parametric
+-- 'Relation'.
+-- Complexity: /O((n + m) * log(n))/ time.
+--
+-- @
+-- gmap f 'empty'      == 'empty'
+-- gmap f ('vertex' x) == 'vertex' (f x)
+-- gmap f ('edge' x y) == 'edge' (f x) (f y)
+-- gmap id             == id
+-- gmap f . gmap g     == gmap (f . g)
+-- @
+gmap :: Ord b => (a -> b) -> SymmetricRelation a -> SymmetricRelation b
+gmap f = SymmetricRelation . R.gmap f . fromSymmetric
+
+-- | Construct the /induced subgraph/ of a given graph by removing the
+-- vertices that do not satisfy a given predicate.
+-- Complexity: /O(m)/ time, assuming that the predicate takes /O(1)/ to
+-- be evaluated.
+--
+-- @
+-- induce ('const' True ) x    == x
+-- induce ('const' False) x    == 'empty'
+-- induce (/= x)               == 'removeVertex' x
+-- induce p . induce q         == induce (\\x -> p x && q x)
+-- 'isSubgraphOf' (induce p x) x == True
+-- @
+induce :: (a -> Bool) -> SymmetricRelation a -> SymmetricRelation a
+induce p = SymmetricRelation . R.induce p . fromSymmetric
+
+-- | Left-to-right /relational composition/ of graphs: vertices @x@ and @z@ are
+-- connected in the resulting graph if there is a vertex @y@, such that @x@ is
+-- connected to @y@ in the first graph, and @y@ is connected to @z@ in the
+-- second graph. There are no isolated vertices in the result. This operation is
+-- associative, has 'empty' and single-'vertex' graphs as /annihilating zeroes/,
+-- and distributes over 'overlay'.
+-- Complexity: /O(n * m * log(m))/ time and /O(n + m)/ memory.
+--
+-- @
+-- compose 'empty'            x                == 'empty'
+-- compose x                'empty'            == 'empty'
+-- compose ('vertex' x)       y                == 'empty'
+-- compose x                ('vertex' y)       == 'empty'
+-- compose x                (compose y z)    == compose (compose x y) z
+-- compose x                ('overlay' y z)    == 'overlay' (compose x y) (compose x z)
+-- compose ('overlay' x y)    z                == 'overlay' (compose x z) (compose y z)
+-- compose ('edge' x y)       ('edge' y z)       == 'edge' x z
+-- compose ('path'    [1..5]) ('path'    [1..5]) == 'edges' [(1,3), (2,4), (3,5)]
+-- compose ('circuit' [1..5]) ('circuit' [1..5]) == 'circuit' [1,3,5,2,4]
```

Comment:
Same here.

Reply:
Corrected to `compose ('circuit' [1..5]) ('circuit' [1..5]) == circuit [1,1,3,3,5,5,2,2,4,4]`

Reply:
That's a nice formulation!

Reply:
Thanks! 


Do you have any idea why is this failling?
![image](https://user-images.githubusercontent.com/20261421/56078715-43534200-5de3-11e9-9e48-8ca5b0f46080.png)

It's really bugging me.
![image](https://user-images.githubusercontent.com/20261421/56078722-582fd580-5de3-11e9-899a-c64a17fce5b1.png)

But:

![image](https://user-images.githubusercontent.com/20261421/56078731-6978e200-5de3-11e9-8a39-13b61b9db6ee.png)

Reply:
Well, I guess the implementation of `compose` is not correct and can produce an inconsistent symmetric relation. That's why we need many tests!

Reply:
Look at the internal representation of `compose (edge 0 0) (edge 0 1)` and you'll see the problem.

Reply:
Ideally, you should add a test for every function in the API that returns a symmetric relation, checking that the result is `consistent`.

Reply:
You are correct! I the `compose (edge 0 0) (edge 0 1)` only has the edge `(0,1)` instead of `[(0,1), (1,0)]`. I already corrected the `compose` function: I was calling the `SymmetricRelation`  constructor instead of the designated `fromRelation` function.

Reply:
Can you point me where to do that? in the `Generic` module?

Reply:
Let's also add the following interesting example and test: `compose x y == compose y x`.

Reply:
What about this one `compose x                (compose y z)      == compose (compose x y) z` ? It does not pass on the tests and I'm having a hard time understanding why.

Reply:
The same for  `compose (edge x y)       (edge y z)       == edge x z` for the case `-1 0 -1` it fails. I'm going to study this and figure what properties should hold

Reply:
But if `SymmetricRelation` is a more restrictive version of `Relation` how come this `compose x (compose y z) == compose (compose x y) z` does not hold? This makes me doubt  about my calculations and the `compose` implementation again...

Reply:
Composition should, of course, be associative, i.e.

`compose x (compose y z) == compose (compose x y) z`

must hold!

If it doesn't, this means there is something wrong with the implementation.

You are right that since this holds for general relations, it should hold for symmetric relations too.

Reply:
> `compose (edge x y) (edge y z) == edge x z` is true if `x /= z`, otherwise `compose (edge x y) (edge y z) == compose (edge y x) (edge z y) == edges [(x,z), (y,y)]`. Should I represent this as properties or should I just remove this one?

Oh, interesting! I think we should have 2-3 less general tests instead:

```haskell
compose (edge 1 x) (edge x 2) == edge 1 2
compose (edge 1 1) (edge 1 1) == edge 1 1
compose (edge 1 2) (edge 2 1) == edges [(1,1), (2,2)]
```

I think this means we need to fix errors in other examples of `compose`, e.g. here:

https://hackage.haskell.org/package/algebraic-graphs-0.3/docs/Algebra-Graph.html#v:compose

Could you please: (1) check that these examples are incorrect in other graph data types? (2) send a separate PR fixing the tests, e.g. by replacing them with the above? If you don't have time for this, don't worry, I'm happy to do this myself.

Reply:
Actually, I think `compose (edge x y) (edge y z) == edge x z` does hold for directed graphs, so it's not as bad as I thought at first :) 

So, let's just switch to these three less general examples:

```haskell
compose (edge 1 x) (edge x 2) == edge 1 2
compose (edge x x) (edge x x) == edge x x
compose (edge 1 x) (edge x 1) == edges [(1,1), (x,x)]
```

Reply:
This is the case because `edge x y` has, internally, the edges [(x,y),(y,x)] so I think that for non Symmetric Graphs, where edge x y only has 1 edge it's fine!

Reply:
Yes, you are right :)

Reply:
So my calculations do not really translate the case where x, y and z are arbitrary instances of a SymmetricRelation ?

Reply:
Sorry, I don't understand what you mean.

Reply:
If you want to keep just a single example, then yes, I think we can use your conditional property.

Reply:
I've been reading about this and if R={(1,2),(2,1),(3,3)} and S={(1,3),(2,2),(3,1)}. Then R∘S={(1,2),(2,3),(3,1)} but S∘R={(1,3),(2,1),(3,2)}.

I'm confused because with this example the composition of 2 `SymmetricRelation` can yield a non `SymmetricRelation`.. currently it does return one because it's running `symmetricClosure` on the result and maybe that's what it's breaking the associative rule. Should I really call `symmetricClosure` on the result of the `compose`? How to deal with this?

Reply:
@snowleopard I'm pretty sure that what breaks the associativity of `compose` in the `SymmetricRelation` module is the fact that it is returned the `symmetricClosure` of the result. For the associativity to be true `symmetricClosure` must only be called on the last call of `compose`...

Relational composition, given R and S is as we expect = `R . S = [(a,c) | (a,b) <- R, (b',c) <- S, b == b']`, but, the `SymmetricRelation`al composition is = `R .. S = (R . S) U (S . R)`, where `(..)` can be seen as the `SymmetricRelation`al composition operator, and R and S are Symmetric.

Knowing this we can try and check if the associativity law holds for this new operator using relational calculus.

Let R, S and T (not the top relation)  be symmetric relations:

`(R .. S) .. T = R .. (S .. T) == { Relational equality (R = S == R ⊆ S /\ S ⊆ R); (..) - definition }`
`( (R . S) U (S . R) ) .. T ⊆ R .. (S .. T) == { (..) - definition }`
`( ( (R . S) U (S . R) ) . T) U (T . ( (R . S) U (S . R) )) ⊆ R .. (S .. T) ==  { U - universal property }`

- `( (R . S) U (S . R) ) . T ⊆ R .. (S .. T)`
- `T . ( (R . S) U (S . R) ) ⊆ R .. (S .. T)`

`== { composition distributes over U }`

- `(R . S . T) U (S . R . T) ⊆ R .. (S .. T)`
- `(T . R . S) U (T . S . R) ⊆ R .. (S .. T)`

`== { U - universal property }`

- `(R . S . T) ⊆ R .. (S .. T)`
- `(S . R . T) ⊆ R .. (S .. T)`
- `(T . R . S) ⊆ R .. (S .. T)`
- `(T . S . R) ⊆ R .. (S .. T)`

`== { did the same for the other side of relational equality}`

- `(R . S . T) ⊆ (R .. S) .. T`
- `(R . T . S) ⊆ (R .. S) .. T`
- `(S . T . R) ⊆ (R .. S) .. T`
- `(T . S . R) ⊆ (R .. S) .. T`

`<= { lowering the upperside x2 }`

- `(R . S . T) ⊆ (R . S . T)`
- `(S . R . T) ⊆ (R . T . S)`
- `(T . R . S) ⊆ (S . T . R)`
- `(T . S . R) ⊆ (T . S . R)`

`== { (S . R . T)° = (T° . R° . S°); (R . T . S)° = (S° . T° . R°); R, S, T are symmetric, so R° = R }`

- `(R . S . T) ⊆ (R . S . T)`
- `(T . R . S) ⊆ (S . T . R)`
- `(T . S . R) ⊆ (T . S . R)`

I derived an impossible equation so I don't think that under the circumstances that I enunciated (calling `symmetricClosure` on `compose`) it's possible to verify the associativity of `SymmetricRelation`al composition.

Note that I understand that composition is associative but the fact that the implementation forces a `symmetricClosure` on every call of `compose` breaks that... And if `symmetricClosure` isn't called then `compose` can give inconsistent instances of `SymmetricRelation`

Reply:
First of all, I'd like to emphasise that `compose` is the "left-to-right relational composition of graphs" as explained in the documentation. The order is opposite compared to the more standard "right-to-left" relational composition. I find left-to-right ordering more natural and convenient to work with.

> I'm confused because with this example the composition of 2 `SymmetricRelation` can yield a non `SymmetricRelation`

Wow, that's pretty cool, I wasn't aware of this. I'll need to remember the 12/21/33-13/22/31 example :) I guess this means we'll need to change the type of the `compose` function so that it returns general `Relation`. We should also add this example as a demonstration that the resulting relation may be non-symmetric. 

Note that if we return a general `Relation`, then we can get rid of the `symmetricClosure` which breaks associativity, as you rightly point out. Thanks for the investigation!

Reply:
We should also remove `compose x y == compose y x` which doesn't hold as your example shows.

Reply:
Just to to be sure I understood your first point `compose x y` is the same as `x . y`, right?

Yes I will change `compose` type but how can we make it so it composes like `compose x . compose y $ z` ?

Reply:
> Just to to be sure I understood your first point `compose x y` is the same as `x . y`, right?

No, I think we actually have `compose x y == y ∘ x` where `∘` is the usual (right-to-left) relational composition.

> Yes I will change compose type but how can we make it so it composes like `compose x . compose y $ z` ?

I think we can't. We'll need to drop the associativity example too, because the types no longer match. It's a bit awkward, but I think it's OK, because having precise types is more important.

> EDIT: Also, regarding changing the type those 3 rules for composition with edge won't apply anymore.

 We can still have something like:

```haskell
compose (edge x y) (edge y z) == if x == z then edges [(x,z), (y,y)] else edge x z
```

Can we?

Reply:
> What about composing a Relation with a SymmetricRelation?

If users need this, they can just apply `fromSymmetric` and compose the result with general relation.

Reply:
Yes, I think that rule is actually more intuitive!

Okay, that makes sense, but I'm having trouble with the API module:
![image](https://user-images.githubusercontent.com/20261421/56083757-4bca6d80-5e21-11e9-973f-0ed22c9f6064.png)

What should I do?

Reply:
I guess you'll need to leave it `undefined` and create custom tests for `compose` in this module.

Reply:
Actually, maybe we shouldn't even provide `compose` in `Symmetric.Relation` module. If users need to compose such relations, they can first convert them to general relations and proceed there. I think this would be preferable, so that we don't have functions named `compose` that have different type signatures.

Reply:
Makes sense to me! I'm going to update the modules with this in mind!